src/registerAllocator/RegisterAllocator.class.ts

```diff
@@ -463,6 +463,9 @@ export class RegisterAllocator {
     const caf = (flagToCheck: AllocationFlags): boolean =>
       ((allocationReq.allocationFlags ?? AllocationFlags.NONE) & flagToCheck) === flagToCheck;
     const inAllocationsTemp = allocationReq.in.map((readVariable) => {
+      const currentLocation = this._allocations[readVariable];
+      if (currentLocation && isRegister(currentLocation.store)) return currentLocation.store;
```

Comment:
why do we need this condition up here? 
I see that for the most part, the remaining parts are checking if it is not a register and then do something. But I am a bit unsure if there is really  no side-effects.

Reply:
If `arg1[3]` is also stored in a register, reading it from the register is more flexible than loading it from memory again. Without this new short-circuit, the next conditional would take the memory path.

Reply:
So instead of
```gas
mov rdx, [rsi]
mulx r8 , r9, [rsi]
```
it will now emit
```gas
mov rdx, [rsi]
mulx r8 , r9, rdx
```
?

Reply:
Then I'd like to write a test case for that.

Reply:
No wait. This would be a square operation, which is handled differently anyway.
This emits
```gas
mov rax, [rsi]
...
mulx r8 , r9, rax
```
instead of
```gas
mov rax, [rsi]
...
mulx r8 , r9, [rsi]
```